### PR TITLE
Do not raise exception when calling set_index on empty dataframe #2819

### PR DESCRIPTION
**Pull request by jessevogt**
Fix for https://github.com/dask/dask/issues/2819

- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Thanks for the PR.

One thing we'll want to consider is how this interacts with [partitions](http://dask.pydata.org/en/latest/dataframe-design.html#partitions)

In this case we end up with 1 partition, `(nan, nan)` for the divisions, and we claim to have known divisions. @jcrist does that sound right to you? I'm not really sure what the correct values should be in this case. I could see an argument for having 0 partitions.

**Comment by jessevogt** [comment]
Appreciate the fast response. My fix is pretty hamfisted. Have been using dask for a little while now but haven't really delved into the internals yet. Any chance you can point me to some examples of a 0 partition DataFrame? Was experimenting around with creating one without much luck.

**Comment by mrocklin** [comment]
For simplicity's sake I would stick to a dataframe with a single partition that evaluates to an empty dataframe.

**Comment by jessevogt** [comment]
@mrocklin @TomAugspurger updated to force repartition when all divisions are null, made sure other types are also handled, and updated the test.

File: dask/dataframe/shuffle.py [hunk 1]
```diff
@@ -64,7 +64,8 @@ def set_index(df, index, npartitions=None, shuffle=None, compute=False,
         divisions, sizes, mins, maxes = base.compute(divisions, sizes, mins, maxes)
         divisions = divisions.tolist()
 
-        if repartition:
+        empty_dataframe_detected = all(pd.isnull(d) for d in divisions)
```
Comment (TomAugspurger) [inline dask/dataframe/shuffle.py]: `pd.isnull` operatres on arrays too, so this could be `pd.isnull(divisions).all()`

**Comment by mrocklin** [comment]
This seems fine to me.  It looks like there is a merge conflict, likely because of your other recently-merged PR :)

**Comment by jessevogt** [comment]
thanks for review. Just rebased master back in, fixed the conflicts and pushed.

**Pull request merged**